src/Language/DifferentialDatalog/DatalogProgram.hs

```diff
@@ -307,3 +311,221 @@ progMirrorInputRelations d prefix =
+                                            (rhsAggExpr r) ] }
+     rAgg = RHSAggregate { rhsVar = varRet,
+                           rhsGroupBy = rhsGroupBy r,
+                           rhsAggFunc = funcName,
+                           rhsAggExpr = input }
+     rCond = RHSCondition { rhsExpr = E ESet { exprPos = nopos,
+                                               exprLVal = E EVarDecl { exprPos = nopos,
+                                                                       exprVName = rhsVar r },
+                                               exprRVal = E ETupField { exprPos = nopos,
+                                                                        exprTuple = E EVar { exprPos = nopos,
+                                                                                             exprVar = varRet },
+                                                                        exprTupField = 1 } } }
+  in [ rAgg, rCond ]
+
+-- OperatorID is based on rule index, rhs index and head index
+-- The 12 MSB are reserved for the rule index.
+-- The next 16 bits are reserved for the rhs index.
+-- The 4 LSB are reserved for the head index.
+generateOperatorIdExpr :: Int -> Int -> Int -> Expr
+generateOperatorIdExpr rlIdx rhsIdx headIdx =
+  let
+    operatorId = toInteger $ (shiftL (rlIdx .&. 4095) 20) .|. (shiftL (rhsIdx .&. 32767) 4) .|. (headIdx .&. 15)
+  in E EBit { exprPos = nopos,
+               exprWidth = 32,
+               exprIVal = operatorId }
+
+ddlogWeightExpr :: Expr
+ddlogWeightExpr =
+  E EVar { exprPos = nopos,
+           exprVar = "ddlog_weight" }
+
+ddlogTimestampExpr :: Expr
+ddlogTimestampExpr =
+  E EVar { exprPos = nopos,
+           exprVar = "ddlog_timestamp" }
+
+generateInspectDebugJoin :: DatalogProgram -> Int -> Rule -> Int -> [RuleRHS]
+generateInspectDebugJoin d ruleIdx rule index =
+  let
+    input1 = (Compile.recordAfterPrefix d rule (index - 1)) !! 0
+    input2 = E EVar { exprPos = nopos,
+                      exprVar = exprVar $ enode $ atomVal $ rhsAtom (ruleRHS rule !! index) }
+    outputs = Compile.recordAfterPrefix d rule index
+  in map (\i -> RHSInspect {rhsInspectExpr = E $ EApply {exprPos = nopos,
+                                                         exprFunc = "debug.debug_event_join",
+                                                         exprArgs = [generateOperatorIdExpr ruleIdx index i,
+                                                                     ddlogWeightExpr,
+                                                                     ddlogTimestampExpr,
+                                                                     input1,
+                                                                     input2,
+                                                                     outputs !! i]}}) [0..length outputs - 1]
+
+generateInspectDebug :: DatalogProgram -> Int -> Rule -> Int -> [RuleRHS]
+generateInspectDebug d ruleIdx rule index =
+  let
+    input1 = if index == 0
```

Comment:
I'm trying to handle the case where the rule only has a single literal.

R5(a, b) :-
  r1 in R1(a, b, _, _).

I want the call to debug_event to have (r1, R5{a, b})

So input is r1, and output is the R5.

If I use recordAfterPrefix  (index), it will return R5. So both input and output will be R5{a, b}. Or does it matter in this case?

Reply:
My bad, your code makes sense!